### DISCUSSION TITLE
Usage of skolem ids and skolem indices

### DISCUSSION THREAD

**Discussion by kfriedberger** [cvc5/cvc5#11331] [Q&A] [answered]
We are currently updating CVC5 in our project ([JavaSMT](https://github.com/sosy-lab/java-smt)) and denoted the new skolemization feature of CVC5.
From a user's point, the skolem ids are not really helpful, because they are some kind of internal solver detail. We would like to avoid skolem ids when printing terms or also when giving them to the user. Is there a way to make skolem ids "transparent" for the user of an API? 

We also found a small regressions in quantifier elimination, maybe also related to skolemization:
We have a simple test `exists y : bv[2]. x * y = 1` (as SMTLIB: `(exists ((y0 (_ BitVec 2))) (= (bvmul x0 y0) #b01))`) with a quantifier-free equivalent `x = 1 OR x = 3` or `extract_0_0 x = 1`. However, CVC5 returns `(= #b01 (bvmul x0 @quantifiers_skolemize_2522))` from `solver.getQuantifierElimination(term)`, which is not equisatisfiable when checked by CVC5. How can we resolve this?

**Comment by ajreynol** [comment]
Currently, there are three places where skolems can appear in our API:
1. Proofs.
2. Lemmas notified to our API Plugin class.
3. Lemmas obtained via the `getUnsatCoreLemmas` command.

They should *not* appear in the results of `getQuantifierElimination`, can you send me the example that triggers this? I can look into a fix.

**Comment by kfriedberger** [comment]
Here is the example from above, using the Java API:
```
    TermManager termManager = new TermManager();
    Solver solver = new Solver(termManager);
    solver.setLogic("ALL");

    int width = 2;

    Term xBv = termManager.mkConst(termManager.mkBitVectorSort(width), "x_bv");
    Term yBv = termManager.mkConst(termManager.mkBitVectorSort(width), "y_bv");
    Term mult = termManager.mkTerm(Kind.BITVECTOR_MULT, xBv, yBv);
    Term body = termManager.mkTerm(Kind.EQUAL, mult, termManager.mkBitVector(width, 1));

    Term xBound = termManager.mkVar(termManager.mkBitVectorSort(width), "y_bv");
    Term quantifiedVars = termManager.mkTerm(Kind.VARIABLE_LIST, xBound);
    Term bodySubst = body.substitute(yBv, xBound);
    Term assertion = termManager.mkTerm(Kind.EXISTS, quantifiedVars, bodySubst);

    Term quantElim = solver.getQuantifierElimination(assertion);
    // --> quantElim is (= #b01 (bvmul x_bv @quantifiers_skolemize_6))
```

**Comment by ajreynol** [comment, marked as answer]
Thanks for the report, I have a fix.

The bug was caused by using "witness terms" from this instantiation procedure: https://homepage.divms.uiowa.edu/~ajreynol/cav18.pdf, where skolems are used to eliminate them. These terms should not be allowed in QE procedures.

The fixed QE is `(or (= #b01 (bvneg x)) (= x #b01))`.

Will send a PR shortly.

**Comment by ajreynol** [comment]
Btw, in the meantime you can use `--no-cegqi-bv` to avoid this issue.

**Comment by daniel-larraz** [comment]
The issue above was resolved by PR #11344.
